grpc/src/main/java/io/stargate/grpc/service/StreamingBatchHandler.java

```diff
@@ -0,0 +1,32 @@
+package io.stargate.grpc.service;
+
+import io.grpc.stub.StreamObserver;
+import io.stargate.db.Persistence;
+import io.stargate.proto.QueryOuterClass;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class StreamingBatchHandler extends BatchHandler {
+  private final AtomicLong inFlight;
+
+  StreamingBatchHandler(
+      QueryOuterClass.Batch batch,
+      Persistence.Connection connection,
+      Persistence persistence,
+      StreamObserver<QueryOuterClass.Response> responseObserver,
+      AtomicLong inFlight,
+      ExceptionHandler exceptionHandler) {
+    super(batch, connection, persistence, responseObserver, exceptionHandler);
+    this.inFlight = inFlight;
+  }
+
+  @Override
+  protected synchronized void setSuccess(QueryOuterClass.Response response) {
```

Comment:
One important find, according to docs https://grpc.github.io/grpc-java/javadoc/io/grpc/stub/StreamObserver.html:  
```
Since individual StreamObservers are not thread-safe, 
if multiple threads will bewriting to a StreamObserver concurrently, 
the application must synchronize calls.
```

We are returning a `StreamObserver` to a caller. All coordination between client and server is done via this one `StreamObserver`. 

Unfortunately, our threading model allows executing `onNext()` on different netty threads. One example output for `Query`:
```
INFO  [Exec Stream Pumper] 2021-12-21 15:55:20,368 ProcessRunner.java:78 - stargate0-0> Executing: RESULT PREPARED cf6d6092b675cffd17f124d83fc57ed8 io.stargate.db.Result$PreparedMetadata@72c1727f (resultMetadata=io.stargate.db.Result$ResultMetadata@60f7c308) idIdempotent=true
INFO  [Exec Stream Pumper] 2021-12-21 15:55:20,381 ProcessRunner.java:78 - stargate0-0> executeWithRetry#setSuccess:  Thread:Native-Transport-Requests-2
INFO  [Exec Stream Pumper] 2021-12-21 15:55:20,381 ProcessRunner.java:78 - stargate0-0> buildResponse: EMPTY RESULT Thread:Native-Transport-Requests-3
INFO  [Exec Stream Pumper] 2021-12-21 15:55:20,381 ProcessRunner.java:78 - stargate0-0> executeWithRetry#setSuccess:  Thread:Native-Transport-Requests-3
```
Note that different threads `Thread:Native-Transport-Requests-2` and `Thread:Native-Transport-Requests-3` are calling `setSuccess` (and therefore `onNext`).

and for `Batch`:
```
INFO  [Exec Stream Pumper] 2021-12-21 16:05:57,741 ProcessRunner.java:78 - stargate0-0> executeWithRetry#setSuccess:  Thread:Native-Transport-Requests-1
INFO  [Exec Stream Pumper] 2021-12-21 16:05:57,742 ProcessRunner.java:78 - stargate0-0> executeWithRetry#setSuccess:  Thread:Native-Transport-Requests-2
INFO  [Exec Stream Pumper] 2021-12-21 16:05:57,742 ProcessRunner.java:78 - stargate0-0> SetSuccess for Batch: Thread: Native-Transport-Requests-1
INFO  [Exec Stream Pumper] 2021-12-21 16:05:57,742 ProcessRunner.java:78 - stargate0-0> SetSuccess for Batch: Thread: Native-Transport-Requests-2
```
I've started noticing this problem because of this error: https://github.com/grpc/grpc-java/issues/4581. 
It manifests only when running `Batch` queries (and `ExecuteBatchStreamingTest`)
I've tried solving this issue by synchronizing `responseObserver.onNext(response);` (in `StreamingBatchHandler` and `StreamingQueryHandler`) but it does not solve the issue. We might need to rethink and re-work our threading model for bi-directional streaming.

Reply:
`streamObserver` is shared across multiple `StreamingBatchHandler`s. Making the method synchronized makes each handler synchronize on itself. This is not enough, you need to use the same object.

This should fix the problem:
```java
synchronized (responseObserver) {
  responseObserver.onNext(response);
}
```

However I'm not a big fan of acquiring a lock in `setSuccess`, because it runs on one of the Persistence's native protocol threads. It's kind of unfortunate that gRPC imposes that on the client application, it makes it hard to write fully async server implementations. I'm not sure if there is a good alternative, at the very least we should evaluate what `onNext` does, to determine how long the thread might be blocked.

Reply:
I've confirmed that this is a problem by changing the 
```
executor = SHARED.newExecutor(1, Integer.MAX_VALUE, "transport", "Native-Transport-Requests");
```
in the `Cassandra311Persistence` - it imposes that all `onNext` are called on the same thread. In this situation, there is no error.

The synchronization on `responseObserver` works, thank you! However, I am not sure about the performance impact of it because as you mentioned, we will block the C* persistence thread. We need to performance test it for sure.

Maybe we could consider changing the threading model for `gRPC`, to make prepare executing on the `grpc-netty-worker` thread, instead of Cassandra native-thread.
Current Context:
1. The incoming gRPC request is executed on the `netty-grpc-thread`
2. At some point, the thread is switched to Cassandra internal thread (native-transport) it's during the `prepare()` method call on persistence
3. after the `prepare()`, all subsequent executions are executed using `native-transport thread`, and not `netty-grpc-thread`.
It’s this method:
```
private CompletionStage<Prepared> maybePrepared(String cql, @Nullable String keyspace) {
    System.out.println("maybePrepared()" + " Thread:" + Thread.currentThread().getName());
    Parameters parameters =
        (keyspace == null)
            ? Parameters.defaults()
            : ImmutableParameters.builder().defaultKeyspace(keyspace).build();

    Prepared preparedInCache = connection.getPrepared(cql, parameters);
    //    CompletableFuture<Prepared> c = new CompletableFuture<>();

    return preparedInCache != null
        ? CompletableFuture.completedFuture(preparedInCache)
        : connection.prepare(cql, parameters); // todo here thread switches, maybe try to switch back?
  }
```
At the beginning of `maybePrepared`, the `netty-thread` is still executing this action. However, `connection.prepare` is returning CompletableFuture that is executed on C* custom thread. We could consider changing `C*prepare()` to not run on the separate executor:
```
public CompletableFuture<Result.Prepared> prepare(String query, Parameters parameters) {
      return executeRequestOnExecutor(
          parameters,
          // The queryStartNanoTime is not used by prepared message, so it doesn't really matter
          // that it's only computed now.
          System.nanoTime(),
          () -> new PrepareMessage(query));
    }

```
(note the `executeRequestOnExecutor`) but on the caller `grpc-netty` thread.
We could also try to get back the thread context (to netty-grpc thread) instead of switching to `native-cassandra`.
However, I am not sure how this can be achieved using `CompletableFuture` API.

Reply:
@tomekl007 I would really like to understand this problem in more details and try to help if possible.

Reply:
OK I think I get it.. 

`2.` won't work you must not block on the `netty-grpc-thread`, and if prepare goes to the data store, boom you are blocked.
`3.` is also not good, because this would decrease the max throughput to the data store and Connection is used by all APIs

So you can effectively do only `1.` Al alternative to synchronized, you can also create a new dedicated executor service with max thread size 1 and do something like

```
CompletableFuture.runAsync(() -> responseObserver.onNext(response), singleThreadedExecutor)
```

this way all interaction with the observer is done on a single thread only. However, not sure if this is more costly than that `synchronized` keyword..

Reply:
Yes, I need firstly to performance test option 1. If it turns out that it is impacting the performance, I will experiment with a dedicated thread pool - but it may also be a bottleneck (even more significant than synchronizing `onNext()`)

Reply:
The all depends on how expensive `onNext()` , is it adding to back of queue or is waiting on a system call. If it's really cheap operation then locking is not a big deal.

Reply:
Yes, there is also no contention between `StreamObeservers` for different clients. The contention is only for a given `StreamObserver` that is processed for a specific caller. So, in the end, the cost of contention might be negligible.